swift/ql/lib/codeql/swift/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -42,47 +43,35 @@ class DataFlowCallable extends TDataFlowCallable {
  * A call. This includes calls from source code, as well as call(back)s
  * inside library callables with a flow summary.
  */
-class DataFlowCall extends TDataFlowCall {
+class DataFlowCall extends ExprNode {
+  DataFlowCall() {
+    this.asExpr() instanceof CallExpr
+  }
+
   /** Gets the enclosing callable. */
   DataFlowCallable getEnclosingCallable() { none() }
-
-  /** Gets a textual representation of this call. */
-  string toString() { none() }
-
-  /** Gets the location of this call. */
-  Location getLocation() { none() }
-
-  /**
-   * Holds if this element is at the specified location.
-   * The location spans column `startcolumn` of line `startline` to
-   * column `endcolumn` of line `endline` in file `filepath`.
-   * For more information, see
-   * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries).
-   */
-  predicate hasLocationInfo(
-    string filepath, int startline, int startcolumn, int endline, int endcolumn
-  ) {
-    this.getLocation().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
-  }
 }
 
 cached
 private module Cached {
   cached
-  newtype TDataFlowCallable = TODO_TDataFlowCallable()
-
-  cached
-  newtype TDataFlowCall = TODO_TDataFlowCall()
+  newtype TDataFlowCallable = TDataFlowFunc(FuncDecl func)
 
   /** Gets a viable run-time target for the call `call`. */
   cached
-  DataFlowCallable viableCallable(DataFlowCall call) { none() }
+  DataFlowCallable viableCallable(DataFlowCall call) { 
+    result = TDataFlowFunc(call.asExpr().(CallExpr).getStaticTarget())
+  }
 
   cached
-  newtype TArgumentPosition = TODO_TArgumentPosition()
+  newtype TArgumentPosition =
+    TThisArgument() or
+    TPositionalArgument(int n) { n in [0 .. 100] } // we rely on default exprs generated in the caller for ordering. TODO: compute range properly. TODO: varargs?
```

Comment:
It looks like the compiler is actually giving us a lot of help here. The following example:
```swift
func sum1(_ n : Int, _ numbers: Int...) {}

sum1(1, 2, 3, 4)
```

has the following AST:
```
(call_expr
  (declref_expr ...) // the function
  (tuple_expr // the arguments
    (integer_literal_expr ... 1 ... ) // the first argument
    (vararg_expansion_expr // the varargs
      (array_expr ...
        (integer_literal_expr ... 2 ... )
        (integer_literal_expr ... 3 ... )
        (integer_literal_expr ... 4 ... )
      )
    )
  )
)
```
So if we just replace `100` with `max(ApplyExpr apply | apply.getNumberOfArguments())` I guess we'll get the correct behavior even for varargs.

An even better solution would be to add an `getIndex()` predicate to the `Argument` class and specify the body of this branch as `n = any(Argument arg).getIndex()`.

Reply:
I pushed a commit with this last suggestion in 2106d487857b8a0e4607e0f4b440a77a0f57cefd.